.github/workflows/main.yml

```
@@ -13,27 +13,130 @@ on:
 
 jobs:
   build:
-    runs-on: ${{ matrix.os }}
-
+    name: ${{ matrix.config.name }}
+    runs-on: ${{ matrix.config.os }}
+    defaults:
+      run:
+        shell: ${{ matrix.config.shell }}
     strategy:
+      fail-fast: false
       matrix:
-        compiler: [clang, gcc]
-        os: [ubuntu-latest]
+        config:
+          - name: Linux GCC
+            os: ubuntu-latest
+            compiler: gcc
+            shell: bash
+
+          - name: macOS Clang
+            os: macos-latest
+            compiler: clang
+            shell: bash
+
+          - name: MSYS2 MINGW64
+            os: windows-latest
+            compiler: gcc
+            shell: 'msys2 {0}'
+            msystem: mingw64
+            msys-env: mingw-w64-x86_64
+
+          - name: MSYS2 MINGW32
+            os: windows-latest
+            compiler: gcc
+            shell: 'msys2 {0}'
+            msystem: mingw32
+            msys-env: mingw-w64-i686
 
     steps:
-      - name: Install dependencies
-        run: sudo apt-get update && sudo apt-get install libfluidsynth-dev libpng-dev libsamplerate0-dev libsdl2-dev libsdl2-image-dev libsdl2-mixer-dev libsdl2-net-dev ninja-build
+      - name: Install pandoc
+        if: runner.os == 'Windows'
+        shell: cmd
+        run: choco install pandoc
+
+      - name: Install dependencies (Linux)
+        if: runner.os == 'Linux'
+        run: |
+          sudo apt-get update
+          sudo apt-get install \
+            libsdl2-dev \
+            libsdl2-mixer-dev \
+            libsdl2-net-dev \
+            libpng-dev \
+            libsamplerate0-dev \
+            libfluidsynth-dev
+
+      - name: Install dependencies (macOS)
+        if: runner.os == 'macOS'
+        env:
+          HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK: 1
+        run: |
+          brew install \
+            automake \
+            pandoc \
+            sdl2 \
+            sdl2_mixer \
+            sdl2_net \
+            libpng \
+            libsamplerate \
+            fluid-synth
+
+      - name: Install dependencies (MSYS2)
+        if: runner.os == 'Windows'
+        uses: msys2/setup-msys2@v2
+        with:
+          msystem: ${{ matrix.config.msystem }}
+          update: false
+          path-type: inherit
+          install: >-
+            autotools
+            zip
+            ${{ matrix.config.msys-env }}-gcc
+            ${{ matrix.config.msys-env }}-SDL2
+            ${{ matrix.config.msys-env }}-SDL2_mixer
+            ${{ matrix.config.msys-env }}-SDL2_net
+            ${{ matrix.config.msys-env }}-libpng
+            ${{ matrix.config.msys-env }}-libsamplerate
+            ${{ matrix.config.msys-env }}-fluidsynth
 
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
```

```suggestion
      - uses: actions/checkout@v4
```
Prefer to use the latest since it's available.